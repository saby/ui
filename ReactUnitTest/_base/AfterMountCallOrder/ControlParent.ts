import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/AfterMountCallOrder/ControlParent';

export default class ControlParent extends Control {
    _template: TemplateFunction = template;
}
