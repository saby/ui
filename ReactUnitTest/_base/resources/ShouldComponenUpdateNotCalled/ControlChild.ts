import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ShouldComponenUpdateNotCalled/ControlChild';

export default class ControlChild extends Control {
    _template: TemplateFunction = template;
    protected _testOption: string;

    protected _beforeMount(options): void {
        this._testOption = options.testOption;
    }

    protected _beforeUpdate(options): void {
        this._testOption = options.testOption;
    }
}
