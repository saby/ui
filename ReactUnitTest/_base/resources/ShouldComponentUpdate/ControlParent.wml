<ReactUnitTest._base.resources.ShouldComponentUpdate.ControlChild
    name="child"
    testOption="{{ _testOption }}"
    timeToMountChild="{{ _timeToMountChild }}"
/>
