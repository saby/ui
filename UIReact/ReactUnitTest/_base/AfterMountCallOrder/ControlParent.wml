<ReactUnitTest._base.AfterMountCallOrder.ControlChild1/>
