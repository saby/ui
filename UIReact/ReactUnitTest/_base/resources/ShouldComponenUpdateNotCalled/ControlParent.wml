<ReactUnitTest._base.resources.ShouldComponenUpdateNotCalled.ControlChild testOption="{{ _testOption }}" />
