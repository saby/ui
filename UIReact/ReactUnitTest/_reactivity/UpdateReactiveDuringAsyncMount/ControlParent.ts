import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_reactivity/UpdateReactiveDuringAsyncMount/ControlParent';

export default class ControlParent extends Control {
    _template: TemplateFunction = template;
    protected _value: string;
}
