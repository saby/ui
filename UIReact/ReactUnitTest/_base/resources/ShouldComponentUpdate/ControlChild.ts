import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ShouldComponentUpdate/ControlChild';

interface IControlChildOptions {
    timeToMountChild: number;
    testOption: { updated: boolean };
}

export default class ControlChild extends Control {
    _template: TemplateFunction = template;
    protected _testOption: { updated: boolean };
    protected _beforeUpdateCallsCount: number = 0;

    protected _beforeMount(options: IControlChildOptions): Promise<void> {
        return new Promise((resolve) => {
            setTimeout(resolve, options.timeToMountChild);
        });
    }

    protected _beforeUpdate(options: IControlChildOptions): void {
        this._testOption = options.testOption;
        this._beforeUpdateCallsCount++;
    }

    checkState(): boolean {
        if (this._testOption?.updated && this._beforeUpdateCallsCount === 1) {
            return true;
        }
    }
}
