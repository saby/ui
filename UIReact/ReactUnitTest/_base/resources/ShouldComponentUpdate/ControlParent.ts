import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ShouldComponentUpdate/ControlParent';
import ControlChild from './ControlChild';

export default class ControlParent extends Control {
    _template: TemplateFunction = template;
    _children: {
        child: ControlChild;
    };
    protected _timeToMountChild: number = ControlParent.timeToMountChild;
    protected _testOption: object = {
        updated: false,
    };

    protected _beforeMount(): Promise<void> | void {
        // eslint-disable-next-line no-unused-expressions, @typescript-eslint/no-unused-expressions
        new Promise((resolve) => {
            setTimeout(() => {
                this._testOption = {
                    updated: true,
                };
                resolve();
            }, ControlParent.timeToChangeOption);
        });
    }

    static readonly timeToMountChild: 30;
    static readonly timeToChangeOption: 15;
}
