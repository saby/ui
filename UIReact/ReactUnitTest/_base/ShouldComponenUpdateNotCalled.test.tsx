/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import ControlParent from './resources/ShouldComponenUpdateNotCalled/ControlParent';
import ControlChild from './resources/ShouldComponenUpdateNotCalled/ControlChild';

describe('Вызов _beforeUpdate даже если не вызвался shouldComponentUpdate', () => {
    let container;

    beforeEach(() => {
        /*
        _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
        Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
         */
        jest.useFakeTimers();
        container = document.createElement('div');
        document.body.appendChild(container);
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
    });

    afterEach(() => {
        jest.useRealTimers();
        unmountComponentAtNode(container);
        container.remove();
        container = null;
        jest.restoreAllMocks();
    });

    it('изменили реактивное свойство родителя, ребенок перестроился', async () => {
        // тут проверяется сценарий, когда родитель и ребенок один раз построились. потом у ребенка что-то поменяли
        // и он перестроился. После этого у родителя меняется реактивное свойство, которое прокинуто в ребенка,
        // но реакт решил не вызывать shouldComponentUpdate, но позовет render для этого контрола.
        // Но необходимо, чтобы _beforeUpdate этого контрола вызвался - последний снепшот упадет, если это не произойдет

        let parent: ControlParent;
        act(() => {
            render(
                <ControlParent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        expect(container).toMatchSnapshot('1. Построили с исходными данными');

        // Делаем так, чтобы первое обновление ребенка прошло как обычно
        await act(async () => {
            parent.changeTestOption('some value');
        });

        // пропустить все таймеры, чтобы произошел полный цикл обновления
        await act(async () => {
            jest.runAllTimers();
        });

        // Мокаем shouldComponentUpdate ребенка после первого вызова, чтобы там не позвался его _beforeUpdate.
        // Ожидаем, что _beforeMount позовется в render и ребенок перестроится
        jest.spyOn(
            ControlChild.prototype,
            'shouldComponentUpdate'
        ).mockImplementation(() => {
            return true;
        });
        await act(async () => {
            parent.changeTestOption('value');
        });

        // пропустить все таймеры, чтобы произошел полный цикл обновления
        await act(async () => {
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot(
            '2. Всё перестроилось с новым состоянием'
        );
    });
});
