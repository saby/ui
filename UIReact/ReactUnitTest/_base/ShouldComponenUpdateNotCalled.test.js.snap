// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Вызов _beforeUpdate даже если не вызвался shouldComponentUpdate изменили реактивное свойство родителя, ребенок перестроился: 1. Построили с исходными данными 1`] = `
<div>
  <div
    tabindex="0"
  />
</div>
`;

exports[`Вызов _beforeUpdate даже если не вызвался shouldComponentUpdate изменили реактивное свойство родителя, ребенок перестроился: 2. Всё перестроилось с новым состоянием 1`] = `
<div>
  <div
    tabindex="0"
  >
    value
  </div>
</div>
`;
